pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -92,85 +87,77 @@ func getRevisionLimits(operatorSpec *operatorv1.StaticPodOperatorSpec) (int32, i
+		}
+		if ns.LastFailedRevision > 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.LastFailedRevision, failedLimit)...)
 		}
 	}
 
-	// Return early if nothing to prune
-	if len(succeededRevisions)+len(failedRevisions)+len(abandonedRevisions) == 0 {
-		klog.V(2).Info("no revision IDs currently eligible to prune")
-		return []int{}, nil
+	if keep.Len() > 0 && keep.List()[0] == 1 && keep.List()[keep.Len()-1] == status.LatestAvailableRevision {
+		return true, nil
 	}
 
-	// Get list of protected IDs
-	protectedSucceededRevisions := protectedRevisions(succeededRevisions, int(succeededRevisionLimit))
-	protectedFailedRevisions := protectedRevisions(failedRevisions, int(failedRevisionLimit))
-	protectedAbandonedRevisions := protectedRevisions(abandonedRevisions, int(abandonedRevisionLimit))
-
-	excludedRevisions := make([]int, 0, len(protectedSucceededRevisions)+len(protectedFailedRevisions)+len(inProgressRevisions)+len(unknownStatusRevisions)+len(protectedAbandonedRevisions))
-	excludedRevisions = append(excludedRevisions, protectedSucceededRevisions...)
-	excludedRevisions = append(excludedRevisions, protectedFailedRevisions...)
-	excludedRevisions = append(excludedRevisions, inProgressRevisions...)
-	excludedRevisions = append(excludedRevisions, unknownStatusRevisions...)
-	excludedRevisions = append(excludedRevisions, protectedAbandonedRevisions...)
-	sort.Ints(excludedRevisions)
-
-	// There should always be at least 1 excluded ID, otherwise we'll delete the current revision
-	if len(excludedRevisions) == 0 {
-		return []int{}, fmt.Errorf("need at least 1 excluded ID for revision pruning")
+	return false, keep
+}
+
+// int32Range returns range of int32 from upper-num+1 to upper.
+func int32RangeBelowOrEqual(upper int32, num int) []int32 {
+	ret := make([]int32, 0, num)
+	for i := 0; i < num; i++ {
+		value := upper - int32(num) + 1 + int32(i)
+		if value > 0 {
+			ret = append(ret, value)
+		}
 	}
-	return excludedRevisions, nil
+	return ret
 }
 
-func (c *PruneController) pruneDiskResources(ctx context.Context, recorder events.Recorder, operatorStatus *operatorv1.StaticPodOperatorStatus, excludedRevisions []int, maxEligibleRevision int) error {
+func (c *PruneController) pruneDiskResources(ctx context.Context, recorder events.Recorder, operatorStatus *operatorv1.StaticPodOperatorStatus, toKeep []int32, maxEligibleRevision int32) error {
 	// Run pruning pod on each node and pin it to that node
 	for _, nodeStatus := range operatorStatus.NodeStatuses {
-		// Use the highest value between CurrentRevision and LastFailedRevision
-		// Because CurrentRevision only updates on successful installs and we still prune on an unsuccessful install
-		if err := c.ensurePrunePod(ctx, recorder, nodeStatus.NodeName, maxEligibleRevision, excludedRevisions, max(nodeStatus.LastFailedRevision, nodeStatus.CurrentRevision)); err != nil {
+		if err := c.ensurePrunePod(ctx, recorder, nodeStatus.NodeName, maxEligibleRevision, toKeep, maxEligibleRevision); err != nil {
```

Comment:
does it mean there is one extra param for `ensurePrunePod` ? (we are passing `maxEligibleRevision` twice)

Reply:
the first is the limit (ignore everything above), and the esecond is the latestAvailableRevision. Preexisting. Cleaned up.